pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -20,33 +20,199 @@ import (
+	ErrNoVSphereInstance = errors.New("no vsphere instance found")
+)
+
+func (nm *NodeManager) DiscoverNode(node *v1.Node) {
+
+	type VmSearch struct {
+		vc string
+		datacenter *vclib.Datacenter
+	}
+
+	var queueChannel chan *VmSearch
+	var wg sync.WaitGroup
+	queueChannel = make(chan *VmSearch, QUEUE_SIZE)
+	nodeUUID := node.Status.NodeInfo.SystemUUID
+
+	go func() {
+		var datacenterObjs []*vclib.Datacenter
+		for vc, vsi := range nm.vsphereInstanceMap {
+			// Create context
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			err := vsi.conn.Connect(ctx)
+			if err != nil {
+				glog.V(4).Info("Discovering node error vc:", err)
+			}
+
+			if vsi.cfg.Datacenters == "" {
+				datacenterObjs, err = vclib.GetAllDatacenter(ctx, vsi.conn)
+				if err != nil {
+					glog.V(4).Info("Discovering node error dc:", err)
+				}
+			} else {
+				datacenters := strings.Split(vsi.cfg.Datacenters, ",")
+				for _, dc := range datacenters {
+					dc = strings.TrimSpace(dc)
+					if dc == "" {
+						continue
+					}
+					datacenterObj, err := vclib.GetDatacenter(ctx, vsi.conn, dc)
+					if err != nil {
+						glog.V(4).Info("Discovering node error dc:", err)
+					}
+					datacenterObjs = append(datacenterObjs, datacenterObj)
+				}
+			}
+
+			for _, datacenterObj := range datacenterObjs {
+				glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
+				queueChannel <- &VmSearch{
+					vc:         vc,
+					datacenter: datacenterObj,
+				}
+			}
+		}
+		close(queueChannel)
+	}()
+
+	for i := 0; i < POOL_SIZE; i++ {
+		go func() {
+			for res:= range queueChannel {
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				vm, err := res.datacenter.GetVMByUUID(ctx, nodeUUID)
+				if vm != nil {
+					glog.V(4).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
+						node.Name, vm, res.vc, res.datacenter.Name())
+
+					nodeInfo := &NodeInfo{dataCenter: res.datacenter, vm: vm, vcServer: res.vc}
+					nm.SetNodeInfo(node.ObjectMeta.Name, nodeInfo)
+					for range queueChannel {}
+					break
+				} else {
+					glog.V(4).Infof("Did not find node %s in vc=%s and datacenter=%s",
+						node.Name, res.vc, res.datacenter.Name(), err)
+				}
+			}
+			wg.Done()
+		}()
+		wg.Add(1)
+	}
+	wg.Wait()
 }
 
-func (nm *NodeManager) registerNode(node *v1.Node) error {
+
+
+func (nm *NodeManager) RegisterNode(node *v1.Node) error {
+	nm.SetNode(node)
+	nm.DiscoverNode(node)
 	return nil
 }
 
-func (nm *NodeManager) unregisterNode(node *v1.Node) error {
+func (nm *NodeManager) UnRegisterNode(node *v1.Node) error {
+	nm.RemoveNode(node)
 	return nil
 }
 
-func (nm *NodeManager) getNodeInfo(nodeName k8stypes.NodeName) (*NodeInfo, error) {
-	return nil, nil
+func (nm *NodeManager) RediscoverNode(nodeName k8stypes.NodeName) error {
+	node, err := nm.GetNode(nodeName)
+
+	if err != nil {
+		return err
+	}
+	nm.DiscoverNode(node)
```

Comment:
If node is moved to a datacenter not part of the list, that is a configuration mistake. Searching on all datacenters is not a good idea and that is the reason why we have the list where the user can limit the search.

Reply:
But then how are you handling scenarios when a node moves out from configured datacenter list?

Reply:
Its a VCP configuration mistake and things would not work.

Reply:
So the admin user has to think forward and put what datacenter the VM might migrate to?
Why don't we search for all datacenter's as it is most unlikely to happen and a one time cost.

Reply:
@abrarshivani @SandeepPissay : Any thoughts?

Reply:
Lets keep the logic simple. 

If the admin does not know where the nodes could be migrated to, the admin can set datastores="". This will make VCP discover nodeVm across all DCs in the VC. If the admin has production and test datacenters in the same VC and wants to limit the scope of the VCP queries to only test datacenter, the admin can set the exact datacenters.